File: Camera/src/main/java/com/example/camera/CameraActivity.java
package com.example.camera;

import android.app.Activity;
import android.hardware.Camera;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.example.camera.preview.MySurfacePreview;
import com.example.camera.util.CameraInstance;
import com.example.camera.util.FaceView;
import com.example.camera.util.GoogleDetectListenerImpl;

import java.lang.ref.WeakReference;

public class CameraActivity extends Activity {

    private MySurfacePreview mySurfacePreview;
    private ImageButton takeBtn;
    private FrameLayout focusLayout;
    private ImageView changeFlashModeIV;
    private ImageView switchCameraIV;
    private RelativeLayout settingRl;
    private FaceView faceView;
    private int width;
    private int height;
    private final MainHandler mainHandler = new MainHandler(this);

    /**
     * Called when the activity was first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        initView();
        bindListeners();
    }

    private void initView() {
        mySurfacePreview = findViewById(R.id.my_surfaceview);
        mySurfacePreview.setHandler(mainHandler);
        takeBtn = findViewById(R.id.take_btn);
        focusLayout = findViewById(R.id.camera_focus_layout);
        int w = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        int h = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        Log.d("showFocusIcon initview", "w " + w + " h " + h);
        focusLayout.measure(w, h);
        width = focusLayout.getMeasuredWidth() / 2;
        height = focusLayout.getMeasuredHeight() / 2;
        Log.d("showFocusIcon initview", "focusLayout.getMeasuredWidth()/2"
                + focusLayout.getMeasuredWidth() / 2 + "focusLayout.getMeasuredHeight()/2"
                + focusLayout.getMeasuredHeight() / 2);
        changeFlashModeIV = findViewById(R.id.flash_iv);
        switchCameraIV = findViewById(R.id.swich_camera_iv);
        settingRl = findViewById(R.id.setting_rl);
        faceView = findViewById(R.id.face_view);
    }

    private void bindListeners() {
        takeBtn.setOnClickListener(v -> CameraInstance.get().doTakePic());
        mySurfacePreview.setOnTouchListener((v, event) -> {
            if (CameraInstance.get().getCameraInfo().facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                return gestureDetector.onTouchEvent(event);
            } else {
                return false;
            }
        });

        changeFlashModeIV.setOnClickListener(v -> CameraInstance.get().setFlashMode(changeFlashModeIV));

        switchCameraIV.setOnClickListener(v -> changeCamera());
    }

    GestureDetector gestureDetector = new GestureDetector(new GestureDetector.SimpleOnGestureListener() {
        @Override
        public boolean onDown(MotionEvent e) {
            Log.d("MyGestureDetector", "onDown");
            return true;
        }

        @Override
        public boolean onSingleTapUp(final MotionEvent e) {
            Log.d("MyGestureDetector", "onSingleTapUp");
            CameraInstance.get().autoFocus((success, camera) -> {
                if (success) {
                    Log.d("renlei", "聚焦成功");
                } else {
                    Log.d("renlei", "聚焦失败");

                }
                focusLayout.setVisibility(View.GONE);
            });
            CameraInstance.get().setFocusArea(CameraActivity.this, e);
            showFocusIcon(e);
            return true;
        }
    });

    private void showFocusIcon(MotionEvent e) {
        int x = (int) e.getX();
        int y = (int) e.getY();
        RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) focusLayout.getLayoutParams();
        lp.leftMargin = (int) (x - width + 0.5);
        lp.topMargin = (int) (y - height + 0.5 + settingRl.getHeight());
        Log.d("showFocusIcon", "x" + x + "y" + y + "params.width" + lp.width
                + "params.height" + lp.height);
        focusLayout.setLayoutParams(lp);
        focusLayout.setVisibility(View.VISIBLE);
    }

    public void changeCamera() {
        CameraInstance.get().doStopPreview();
        int newCameraId = (CameraInstance.get().getCameraId() + 1) % 2;
        CameraInstance.get().doOpenCamera(newCameraId);
        CameraInstance.get().doStartPreview(mySurfacePreview.getHolder());
        if (newCameraId == Camera.CameraInfo.CAMERA_FACING_BACK) {
            switchCameraIV.setImageResource(R.drawable.camera_setting_switch_back);
            changeFlashModeIV.setVisibility(View.VISIBLE);
        } else {
            switchCameraIV.setImageResource(R.drawable.camera_setting_switch_front);
            changeFlashModeIV.setVisibility(View.GONE);
        }
    }

    private void startGoogleDetect() {
        Camera.Parameters parameters = CameraInstance.get().getCameraParameters();
        Camera camera = CameraInstance.get().getCamera();
        if (parameters.getMaxNumDetectedFaces() > 0) {
            if (faceView != null) {
                faceView.clearFaces();
                faceView.setVisibility(View.VISIBLE);
            }
            camera.setFaceDetectionListener(new GoogleDetectListenerImpl(mainHandler));
            camera.startFaceDetection();
        }
    }

    private static class MainHandler extends Handler {

        private final WeakReference<CameraActivity> mHost;

        MainHandler(CameraActivity host) {
            mHost = new WeakReference<>(host);
        }

        @Override
        public void handleMessage(final Message msg) {
            CameraActivity host = mHost.get();
            if (host == null) return;
            switch (msg.what) {
                case CameraInstance.PREVIEW_HAS_STARTED:
                    host.startGoogleDetect();
                    Log.e("renlei110", "开启人脸识别");
                    break;
                case CameraInstance.RECEIVE_FACE_MSG:
                    host.faceView.setFaces((Camera.Face[]) msg.obj);
                    Log.e("renlei111", "收到人脸识别的信息");
                    break;
            }
        }
    }
}

File: Camera/src/main/java/com/example/camera/CameraActivity.kt
package com.example.camera

import android.Manifest
import android.graphics.PixelFormat
import android.hardware.Camera
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import com.example.camera.databinding.MainBinding

class CameraActivity : AppCompatActivity() {

    private var binding: MainBinding? = null
    private var width = 0
    private var height = 0

    /**
     * Called when the activity was first created.
     */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        supportActionBar?.hide()
        binding = MainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        initView()
        bindListeners()
    }

    override fun onStart() {
        super.onStart()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return
        requestPermissions(arrayOf(Manifest.permission.CAMERA), 0x10)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != 0x10) return
        Log.d(
            TAG,
            "onRequestPermissionsResult() permissions = ${permissions.contentToString()}, grantResults = ${grantResults.contentToString()}"
        )
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
    }

    private fun initView() {
        //translucent半透明 transparent透明
        binding!!.surfaceView.holder.setFormat(PixelFormat.TRANSPARENT)
        binding!!.surfaceView.holder.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(holder: SurfaceHolder) {
                CameraWrapper.get().open()
            }

            override fun surfaceChanged(
                holder: SurfaceHolder,
                format: Int,
                width: Int,
                height: Int
            ) {
                Log.d(
                    TAG,
                    "surfaceChanged() called with: format = $format, width = $width, height = $height"
                )
                CameraWrapper.get().preview(holder)
            }

            override fun surfaceDestroyed(holder: SurfaceHolder) {
                CameraWrapper.get().stopPreview()
            }
        })

        binding!!.cameraFocusLayout.post {
            width = binding!!.cameraFocusLayout.width / 2
            height = binding!!.cameraFocusLayout.height / 2
        }
    }

    private fun bindListeners() {
        binding!!.takeBtn.setOnClickListener { v: View? -> CameraWrapper.get().doTakePic() }
        binding!!.surfaceView.setOnTouchListener { _: View?, event: MotionEvent? ->
            if (CameraWrapper.get().cameraInfo.facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                return@setOnTouchListener gestureDetector.onTouchEvent(event)
            } else {
                return@setOnTouchListener false
            }
        }
        binding!!.flashIv.setOnClickListener {
            CameraWrapper.get().setFlashMode(binding!!.flashIv)
        }
        binding!!.swichCameraIv.setOnClickListener { changeCamera() }
    }

    private var gestureDetector =
        GestureDetector(object : GestureDetector.SimpleOnGestureListener() {
            override fun onDown(e: MotionEvent): Boolean {
                Log.d("MyGestureDetector", "onDown")
                return true
            }

            override fun onSingleTapUp(e: MotionEvent): Boolean {
                Log.d("MyGestureDetector", "onSingleTapUp")
                CameraWrapper.get().autoFocus { success: Boolean, _: Camera? ->
                    if (success) {
                        Log.d(TAG, "聚焦成功")
                    } else {
                        Log.d(TAG, "聚焦失败")
                    }
                    binding!!.cameraFocusLayout.visibility = View.GONE
                }
                CameraWrapper.get().setFocusArea(this@CameraActivity, e)
                showFocusIcon(e)
                return true
            }
        })

    private fun showFocusIcon(e: MotionEvent) {
        val x = e.x.toInt()
        val y = e.y.toInt()
        val lp = binding!!.cameraFocusLayout.layoutParams as FrameLayout.LayoutParams
        lp.leftMargin = (x - width + 0.5).toInt()
        lp.topMargin = (y - height + 0.5 + binding!!.swichCameraIv.height).toInt()
        Log.d(
            TAG, "showFocusIcon x" + x + "y" + y + "params.width" + lp.width
                    + "params.height" + lp.height
        )
        binding!!.cameraFocusLayout.layoutParams = lp
        binding!!.cameraFocusLayout.visibility = View.VISIBLE
    }

    private fun changeCamera() {
        CameraWrapper.get().stopPreview()
        val newCameraId: Int = (CameraWrapper.get().cameraId + 1) % 2
        CameraWrapper.get().open(newCameraId)
        CameraWrapper.get().preview(binding!!.surfaceView.holder)
        if (newCameraId == Camera.CameraInfo.CAMERA_FACING_BACK) {
            binding!!.swichCameraIv.setImageResource(R.drawable.camera_setting_switch_back)
        } else {
            binding!!.swichCameraIv.setImageResource(R.drawable.camera_setting_switch_front)
        }
    }

    companion object {
        private const val TAG = "CameraActivity"
    }
}